### DISCUSSION TITLE
add_raster doesn't work with ssh remote environment

### DISCUSSION THREAD

**Discussion by zxdawn** [opengeos/leafmap#469] [Q&A] [answered]
As discussed [here](https://github.com/banesullivan/localtileserver/issues/158), the localtileserver needs some special settings to make `add_raster` function in leafmap work with SSH remote environment. I will update the answer when we get the best answer.

**Comment by zxdawn** [comment, marked as answer]
If you are running leafmap remotely, you need to install `jupyter-server-proxy` and add `export LOCALTILESERVER_CLIENT_PREFIX='proxy/{port}'` to your remote environment, then `add_raster` will work well.

**Comment by giswqs** [comment]
Thank you for sharing the solution.

**Reply by zxdawn** [reply]
@giswqs You're welcome. Actually I realize this method works in Remote Jupyter Notebook, but not for Remote Streamlit ...

```
import os
os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = 'proxy/{port}/'

import leafmap.foliumap as leafmap
import streamlit as st

url = 'https://github.com/opengeos/data/raw/main/raster/srtm90.tif'

image = 'srtm90.tif'
if not os.path.exists(image):
    leafmap.download_file(url, image)

filename = 'srtm90.tif'

# plot the basemap
m = leafmap.Map()
m.add_basemap('Esri.WorldImagery')

m.add_raster(filename, band=1, palette='plasma', layer_name='test')

m.to_streamlit()
```

<img width="1453" alt="image" src="https://github.com/opengeos/leafmap/assets/30388627/738b8b10-d889-4896-8723-03b219104be0">

**Reply by zxdawn** [reply]
This line `os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = 'proxy/{port}/'` will cause the issue even with local streamlit. I suppose the port of streamlit is special.

**Comment by giswqs** [comment]
Streamlit does not support localtileserver. Don't wait your time on this. See https://github.com/streamlit/streamlit/pull/4677

The workaround to to make the raster available through HTTP, then use `m.add_cog_layer()` to display it.

**Reply by zxdawn** [reply]
Ha, strange. If I comment out `os.environ['LOCALTILESERVER_CLIENT_PREFIX'] = 'proxy/{port}/'`, `add_raster` works with local file. Didn't it use localtileserver?

<img width="1493" alt="image" src="https://github.com/opengeos/leafmap/assets/30388627/c0c91f19-b772-44cf-992a-0f9ecbbf6586">

**Reply by giswqs** [reply]
It will only work if you use it locally. If you deploy it to the Streamlt Cloud, it will not work.

**Reply by zxdawn** [reply]
Got it. Actually, I only want to run it locally on my Linux machine and connect to it using SSH from laptop (because the data is private and large). It seems tricky ...

**Reply by giswqs** [reply]
Try Solara: https://giswqs-solara-geospatial.hf.space/leafmap

**Reply by zxdawn** [reply]
Thanks. I will try that instead of streamlit!

**Reply by rh-afk** [reply]
Hi @giswqs , I just try your solara geospatial, unfortunately there was an error:
![image](https://github.com/opengeos/leafmap/assets/141318843/f2003c24-780e-402f-a92b-6782a60c9153)
I  have restart the kernel, shutdown the kernel or even with shutdown server in jupyter hub and up again the server. However there still have error. did you have any suggestion.
If run individual leaflet(without solara) there is no issue
